sdk/src/main/java/com/google/cloud/dataflow/sdk/util/common/NameContext.java

```diff
@@ -0,0 +1,37 @@
+/*******************************************************************************
+ * Copyright (C) 2016 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ ******************************************************************************/
+package com.google.cloud.dataflow.sdk.util.common;
+
+import com.google.auto.value.AutoValue;
+
+/**
+ * A NameContext provides a way for counters to control what context they will be associated with in
+ * monitoring.
+ */
+@AutoValue
+public abstract class NameContext {
+  /** Create a NameContext with both an originalName and systemName. */
```

Comment:
`{@link NameContext}`, `{@code originalName}` etc.

Also -- what if we wanted counters that weren't associated with a step (eg., that was pipeline wide). Would that be a different kind of `NameContext`?